---
title: "RedaCT-Seq"
author: "Dave"
date: "`r Sys.Date()`"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library('data.table')
library('dplyr')
```

```{r eval=T}

####################################
# Column order for the data:
####################################

# WT NaBH4
# KO NaBH4
# WT Untreated control

####################################
# Process all mismatches
####################################
pileup <- fread('sampledata/mpileup_output_chr19_min10_parsed.txt')
nrow(pileup)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Optional filter 
## Reduce to locations on primary assembly 
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
mychrs <- c(paste0("chr",seq(1,22,1)),"chrX")
pileup <- pileup %>% filter(chr %in% mychrs)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# Get mismatch percentages
All.ref <- data.frame(pileup[, c(1:13)], mis_WT_NaBH4=with(pileup, (Sub_0)/(depth_0)), mis_KO_NaBH4=with(pileup, (Sub_1)/(depth_1)), mis_WT_Ctrl=with(pileup, (Sub_2)/(depth_2)),mis_WT_Ctrl_total=with(pileup, ((depth_2) - (Ref_2))/(depth_2)))

#Get ratio between samples
All.ref <- data.frame(All.ref[, c(1:17)], WT_KO_NaBH4_ratio=with(All.ref, mis_WT_NaBH4/mis_KO_NaBH4), WT_WT_Ctrl_ratio=with(All.ref, mis_WT_NaBH4/mis_WT_Ctrl), KO_WT_NaBH4_ratio=with(All.ref, mis_KO_NaBH4/mis_WT_NaBH4))

head(All.ref)
dim(All.ref)

# Preserve the unfiltered object under
# a different name so you can go back to it.
# All.ref.unfiltered <- All.ref

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Apply some filters
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

# Depth at least 10 in WT AND KO AND Untreated
All.ref <- All.ref %>% filter(depth_0 >= 10 & depth_1 >= 10 & depth_2 >= 10)
dim(All.ref)

#Filter coverage in Substitute  base for > 4 in either treated sample
All.ref <- All.ref %>% filter(Sub_0 >= 4 | Sub_1 >= 4)
dim(All.ref)

#Filter by mismatch fraction >= 1% in either treated sample
All.ref <- All.ref %>% filter(mis_WT_NaBH4 >= 0.01 | mis_KO_NaBH4 >= 0.01)
dim(All.ref)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Obtain ac4C candidate sites ( in balanced fashion (KO and WT))
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Combined.ref.allmm <- All.ref

#~~~~~~~~~~~~~~~~~~~~~~~~
# Append consistent identifiers
#~~~~~~~~~~~~~~~~~~~~~~~~

Combined.ref.allmm$locID <- paste0(Combined.ref.allmm$chr,":",Combined.ref.allmm$loc)
Combined.ref.allmm$subID <- paste0(Combined.ref.allmm$chr,":",Combined.ref.allmm$loc,":",Combined.ref.allmm$ref,">",Combined.ref.allmm$sub)

#~~~~~~~~~~~~~~~~~~~~~~~~
# Save objects (optional)
#~~~~~~~~~~~~~~~~~~~~~~~~

#dim(Combined.ref.allmm)
#save(Combined.ref.allmm, file="Combined.ref.allmm.Rdata")

# These are optional to save for diagnostics
#save(All.ref.unfiltered, file="Robjects/All.ref.unfiltered.Rdata")
#save(Combined.ref.CT, file="Robjects/Combined.ref.CT_pooled.Rdata")
```

# Diagnostics
## Exploring mismatch profiels before resolving stand 
## or performing any filtering

```{r volcano_allmm_wtsites_diagnostics, fig.width=10, fig.height=8}
Combined.ref.allmm$Fisher_WT_KO_BH4 = NA
for(i in 1:length(Combined.ref.allmm$chr)){
  Combined.ref.allmm$Fisher_WT_KO_BH4[i] = fisher.test(matrix(c(Combined.ref.allmm$Sub_0[i],Combined.ref.allmm$Ref_0[i],Combined.ref.allmm$Sub_1[i],Combined.ref.allmm$Ref_1[i]), nrow=2))$p.value
}

Combined.ref.allmm$genomic_conversion <- paste0(Combined.ref.allmm$ref,">",Combined.ref.allmm$sub)

conv_tab <- table(Combined.ref.allmm$genomic_conversion)
conv_tab[order(conv_tab)]

myconvs <- names(conv_tab)

# Set cutoff

myp <- 0.01
myratio <- 5

op <- par(mfrow = c(3,4))
for (i in 1:length(myconvs)) {
  
  myconv <- myconvs[i]
  myset <- Combined.ref.allmm[Combined.ref.allmm$genomic_conversion %in% myconv,]
  
  
  myx <- log2(myset$WT_KO_NaBH4_ratio)
	myy <- -log10(myset$Fisher_WT_KO_BH4)
	myy[myy >= 10] <- 10
	myx[myx >= 6] <- 6
	myy[myy <= -10] <- -10
	myx[myx <= -6] <- -6

	# Get the sigs
	mysig.wt <- myset[myset$Fisher_WT_KO_BH4 <= myp & myset$WT_KO_NaBH4_ratio >= myratio,]
	#mysig.wt.ct <- mysig.wt[mysig.wt$conversion == "C>T",]
	myx.wt <- log2(mysig.wt$WT_KO_NaBH4_ratio)
	myy.wt <- -log10(mysig.wt$Fisher_WT_KO_BH4)

	#mysig.ko <- myset[myset$Fisher_WT_KO_BH4 <= myp & myset$WT_KO_NaBH4_ratio <= (1/myratio),]
	mysig.ko <- myset[myset$Fisher_WT_KO_BH4 <= myp & myset$KO_WT_NaBH4_ratio >= myratio,]
	#mysig.ko.ct <- mysig.ko[mysig.ko$conversion == "C>T",]
	myx.ko <- log2(mysig.ko$WT_KO_NaBH4_ratio)
	myy.ko <- -log10(mysig.ko$Fisher_WT_KO_BH4)

	myy.wt[myy.wt >= 10] <- 10
	myx.wt[myx.wt >= 6] <- 6
	myy.wt[myy.wt <= -10] <- -10
	myx.wt[myx.wt <= -6] <- -6

	myy.ko[myy.ko >= 10] <- 10
	myx.ko[myx.ko >= 6] <- 6
	myy.ko[myy.ko <= -10] <- -10
	myx.ko[myx.ko <= -6] <- -6  
	#~~~~~~~~~~~~~~~~~~~~~
	# Color coding by sig
	#~~~~~~~~~~~~~~~~~~~~~
	plot(myx,myy,main=paste0(myconv,
	" conversion (raw genomic calls)\n(WT vs NAT10-/-)"),xlab="FC",ylab="-log10(P unadj)",col="grey40",las=1,xlim=c(-6,6),ylim=c(0,10),pch=19)
	points(myx.wt,myy.wt,col="red",pch=19)
	points(myx.ko,myy.ko,col="blue",pch=19)
	abline(v=0,col="red",lwd=1)
	abline(v=log2(myratio),col="red",lwd=1,lty=2)   
	abline(v=log2(1/myratio),col="red",lwd=1,lty=2) 
	abline(h=-log10(myp),col="red",lwd=1)   
	legend("top",c(as.character(length(myx.wt)),as.character(length(myx.ko))),pch=19,col=c("red","blue"))
}
par(op)

```


# Librarys for conversion for tx space
```{r, eval=T}
library('rtracklayer')
library('GenomicFeatures')
library('genomation')
#library('readxl')
library('stringr')
```


```{r, eval=T}

#Genomic_candidates <- Combined.ref.allmm[Combined.ref.allmm$Fisher_WT_KO_BH4 <= 0.05,]
mysites <- Combined.ref.allmm
mybed <- data.frame(chr = mysites$chr,start = format(mysites$loc - 1,scientific=F) ,end = format(mysites$loc,scientific=F),id = mysites$subID,score = 1,strand = "*")
mybed$start <- str_trim(mybed$start, side = "both")
mybed$end <- str_trim(mybed$end, side = "both")
write.table(mybed,file="genomicsites.bed",col.names=F,quote=F,row.names=F,sep="\t")
```

```{r MapToTx_set1, eval=T}


NaBH4sites <- Combined.ref.allmm
mybed <- genomation::readBed("genomicsites.bed")

#mybed <- genomation::readBed("BED/genomicsites_allmm_pool_1percent_KO_birectional_balanced.bed",track.line=FALSE)
# Note I do this (write out bed, reload in bed) because its the only way I know to get in genomation format

dim(mybed)

mytxcanon <- rtracklayer::import("sampledata/genes_chr19.gtf")
myTxDbcanon <- makeTxDbFromGRanges(mytxcanon)

exon_by_tx_canon <- exonsBy(myTxDbcanon, by="tx", use.names=TRUE)
mapped_exon_canon <- mapToTranscripts(mybed, exon_by_tx_canon,ignore.strand=FALSE)


#~~~~~~~~~~~~~~~~~~
# Tabulate feature lengths
#~~~~~~~~~~~~~~~~~

transcriptInfo <- transcriptLengths(myTxDbcanon, with.cds_len=TRUE,
                  with.utr5_len=TRUE, with.utr3_len=TRUE)


gr <-  mapped_exon_canon       
query.idx <- mcols(mapped_exon_canon)[,1] 
query.ids <- mcols(mybed)[,2]
query.sites <- query.ids[query.idx]
length(query.sites)

df <- data.frame(seqnames=seqnames(gr),
  starts=start(gr)-1,
  ends=end(gr),
  names=c(rep(".", length(gr))),
  scores=c(rep(".", length(gr))),
  strand=strand(gr),
  subID = query.sites)


#~~~~~~~~~~~~~~
# Adding annotation, feature location
#~~~~~~~~~~~~~~
        
df$txlength <- transcriptInfo$tx_len[match(df$seqnames,transcriptInfo$gene_id)]        
df$UTR5length <- transcriptInfo$utr5_len[match(df$seqnames,transcriptInfo$gene_id)]        
df$CDSlength <- transcriptInfo$cds_len[match(df$seqnames,transcriptInfo$gene_id)]        
df$UTR3length <- transcriptInfo$utr3_len[match(df$seqnames,transcriptInfo$gene_id)]        

myfeatures <- NULL
myfeatureposs <- NULL

for (i in 1:nrow(df)) {

        myrow <- df[i,]
        mypos <- myrow$starts
        if (myrow$CDSlength < 1) {
          myfeature = "none"
          myfeatures <- c(myfeatures,myfeature)
          myfeatureposs <- c(myfeatureposs,mypos)
        } else {
					if (mypos <= myrow$UTR5length) { myfeature = "UTR5" }
					if (mypos > myrow$UTR5length & mypos <= (myrow$UTR5length + myrow$CDSlength)) { myfeature = "CDS" }
					if (mypos > (myrow$UTR5length + myrow$CDSlength)) { myfeature = "UTR3" }      
					myfeatures <- c(myfeatures,myfeature)
					if (myfeature == "UTR5") { myfeaturepos <- mypos }        
					if (myfeature == "CDS") { myfeaturepos <- mypos - myrow$UTR5length }        
					if (myfeature == "UTR3") { myfeaturepos <- mypos - myrow$UTR5length - myrow$CDSlength }           
					myfeatureposs <- c(myfeatureposs,myfeaturepos)
				}
}
 
df$feature <- myfeatures
df$feature_pos <- myfeatureposs

table(df$feature)
prop.table(table(df$feature))

df$pos_rel_aTIS <- df$starts - df$UTR5length
df$pos_rel_aTIS[df$feature == "none"] <- NA
df$feature_pos[df$feature == "none"] <- NA
#~~~~~~~~~~~~~~~~
# Correct the strand etc
#~~~~~~~~~~~~~~~~~

NaBH4_join <- left_join(df,NaBH4sites,by="subID")

NaBH4_join$ref_strandcorrect <- NaBH4_join$ref
NaBH4_join$sub_strandcorrect <- NaBH4_join$sub

NaBH4_join$ref_strandcorrect[NaBH4_join$strand == "-" & NaBH4_join$ref == "C"] <- "G"
NaBH4_join$ref_strandcorrect[NaBH4_join$strand == "-" & NaBH4_join$ref == "G"] <- "C"
NaBH4_join$ref_strandcorrect[NaBH4_join$strand == "-" & NaBH4_join$ref == "A"] <- "T"
NaBH4_join$ref_strandcorrect[NaBH4_join$strand == "-" & NaBH4_join$ref == "T"] <- "A"

NaBH4_join$sub_strandcorrect[NaBH4_join$strand == "-" & NaBH4_join$sub == "C"] <- "G"
NaBH4_join$sub_strandcorrect[NaBH4_join$strand == "-" & NaBH4_join$sub == "G"] <- "C"
NaBH4_join$sub_strandcorrect[NaBH4_join$strand == "-" & NaBH4_join$sub == "A"] <- "T"
NaBH4_join$sub_strandcorrect[NaBH4_join$strand == "-" & NaBH4_join$sub == "T"] <- "A"

NaBH4_join$conversion <- paste0(NaBH4_join$ref_strandcorrect,">",NaBH4_join$sub_strandcorrect)

prop.table(table(NaBH4_join$conversion[NaBH4_join$strand == "+"])) * 100
prop.table(table(NaBH4_join$conversion[NaBH4_join$strand == "-"])) * 100

table(NaBH4_join$conversion)
prop.table(table(NaBH4_join$conversion))

#~~~~~~~~~~~~~~~~~~~~~~~
# How many have ambiguous mapping to transcripts?
#~~~~~~~~~~~~~~~~~~~~~~

maptab <- table(NaBH4_join$subID)
ambiguous_map <- names(maptab[maptab > 1])
length(ambiguous_map)
table(NaBH4_join$conversion[NaBH4_join$subID %in% ambiguous_map])

#~~~~~~~~~~~~~~~~~~~~~~~
# How many have more than one substitution type?
#~~~~~~~~~~~~~~~~~~~~~~

# Get the table of unique subID and locID pairs
# Find locID (genomic locations) that > 1 subID (substitutions)

testtab <- unique(NaBH4_join[,c("subID","locID")])
multiple_sub_tab <- table(testtab$locID)
multiple_subs <- names(multiple_sub_tab[multiple_sub_tab > 1])


#~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~
# Filter ambiguous mapping or mulitple subs
#~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~

NaBH4_join_filter <- NaBH4_join %>% filter(!(subID %in% ambiguous_map))
NaBH4_join_filter <- NaBH4_join_filter %>% filter(!(locID %in% multiple_subs))

# testing filter
table(NaBH4_join$subID %in% ambiguous_map)
table(NaBH4_join$locID %in% multiple_subs)


table(NaBH4_join_filter$conversion)
prop.table(table(NaBH4_join_filter$conversion))
barplot(table(NaBH4_join_filter$conversion),las=2,main="Substitutions in unfiltered sites")



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Save the results
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
dim(NaBH4_join_filter)
dim(NaBH4_join)
NaBH4_join_allmm_pool <- NaBH4_join_filter
dim(NaBH4_join_allmm_pool)
save(NaBH4_join_allmm_pool,file="NaBH4_join_allmm_txmapped.Rdata")    






```

# Load the object for prototyping below
```{r eval=F}
print(load(file="NaBH4_join_allmm_txmapped.Rdata"))
```

# Filter and get candidates

```{r}
#~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Get data for just 1.25%
#~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~

#~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Get candidates that pass comparison to untreated
#~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~

#`````````````````````````````
# Increase minimum depth for comparison to untreated
# Coverage in Untreated >= 25
#`````````````````````````````

mycandidates <- NaBH4_join_allmm_pool[NaBH4_join_allmm_pool$depth_2 >= 25 & !(is.na(NaBH4_join_allmm_pool$depth_2)),]


#~~~~~~~~~~~~~~
# Add more fields for diagnostics
#~~~~~~~~~~~~~~
mycandidates$diffMM <- mycandidates$mis_WT_NaBH4 - mycandidates$mis_KO_NaBH4
mycandidates$mean_NaBH4_depth <- (mycandidates$depth_0 + mycandidates$depth_1) / 2
mycandidates$max_NaBH4_MM <- apply(mycandidates[,c("mis_KO_NaBH4","mis_WT_NaBH4")],1,max)
mycandidates$Perc_max_MM_lost <- (mycandidates$max_NaBH4_MM - mycandidates$mis_WT_Ctrl)/mycandidates$max_NaBH4_MM

#`````````````````````````````
# 90% mismatches lost
# This is balanced for WT/KO - takes the maximum of either
#`````````````````````````````
prop.table(table(mycandidates$Perc_max_MM_lost > 0.90))
mycandidates <- mycandidates[mycandidates$Perc_max_MM_lost >= 0.9,]

#`````````````````````````````
# Total mismatches at this site < 1% in WT untreated
#`````````````````````````````
mycandidates <- mycandidates[mycandidates$mis_WT_Ctrl_total < 0.01 ,]

#~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~
# Final selection for testing 
# Raising threshold to 1.25%
#~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~

# Three sets here:
#1 Prefiltered (before anything is applied)
#2 Balanced (Applying the same threshold to WT and KO)
#3 1.25% in WT (As presented in the manuscript)

#prefiltered_candidates <- mycandidates

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Restrict to minimum 1.25% MM in either treated sample
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

balanced_candidates <- mycandidates[mycandidates$mis_WT_NaBH4 >= 0.0125 | mycandidates$mis_KO_NaBH4 >= 0.0125,]


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Perform FET
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

balanced_candidates$Fisher_WT_KO_BH4 = NA
for(i in 1:nrow(balanced_candidates)){
  balanced_candidates$Fisher_WT_KO_BH4[i] = fisher.test(matrix(c(balanced_candidates$Sub_0[i],balanced_candidates$Ref_0[i],balanced_candidates$Sub_1[i],balanced_candidates$Ref_1[i]), nrow=2))$p.value
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# FDR correct
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

balanced_candidates$Fisher_WT_KO_BH4.adj <- p.adjust(balanced_candidates$Fisher_WT_KO_BH4,method="BH")




```


# Volcano 
# (1.25% in either WT or KO, FDR adjusted)


```{r volcano_CT_wt, fig.width=5, fig.height=5}


candidates <- balanced_candidates

#~~~~~~~~~~~~~~~~~~~~~
# Now filter with WT / NAT10 comparison
#~~~~~~~~~~~~~~~~~~~~~
# With re-FDR adjusted
myp <- 0.05
myratio <- 5
mysig_allmm <- candidates[candidates$Fisher_WT_KO_BH4.readj <= myp & candidates$WT_KO_NaBH4_ratio >= myratio,]
dim(mysig_allmm)
table(mysig_allmm$conversion)
#mysig <- evaluateCutoffsAdj(readj,myp,myratio)

table(mysig_allmm$conversion)
prop.table(table(mysig_allmm$conversion))

myset <- candidates[candidates$conversion == "C>T",]

	myx <- log2(myset$WT_KO_NaBH4_ratio)
	myy <- -log10(myset$Fisher_WT_KO_BH4.adj)
	myy[myy >= 5] <- 5
	myx[myx >= 6] <- 6
	myy[myy <= -5] <- -5
	myx[myx <= -6] <- -6
	
  # Get the sigs
	mysig.wt <- myset[myset$Fisher_WT_KO_BH4.adj <= myp & myset$WT_KO_NaBH4_ratio >= myratio,]
  #mysig.wt.ct <- mysig.wt[mysig.wt$conversion == "C>T",]
	myx.wt <- log2(mysig.wt$WT_KO_NaBH4_ratio)
  myy.wt <- -log10(mysig.wt$Fisher_WT_KO_BH4.adj)
  
	mysig.ko <- myset[myset$Fisher_WT_KO_BH4.adj <= myp & myset$WT_KO_NaBH4_ratio <= (1/myratio),]
	#mysig.ko.ct <- mysig.ko[mysig.ko$conversion == "C>T",]
  myx.ko <- log2(mysig.ko$WT_KO_NaBH4_ratio)
  myy.ko <- -log10(mysig.ko$Fisher_WT_KO_BH4.adj)

	myy.wt[myy.wt >= 5] <- 5
	myx.wt[myx.wt >= 6] <- 6
	myy.wt[myy.wt <= -5] <- -5
	myx.wt[myx.wt <= -6] <- -6
	
	myy.ko[myy.ko >= 5] <- 5
	myx.ko[myx.ko >= 6] <- 6
	myy.ko[myy.ko <= -5] <- -5
	myx.ko[myx.ko <= -6] <- -6  
#~~~~~~~~~~~~~~~~~~~~~
# Color coding by sig
#~~~~~~~~~~~~~~~~~~~~~
plot(myx,myy,main=paste0(
  "C>T conversion rate \ndifferences (WT vs NAT10-/-)"),xlab="FC",ylab="-log10(P adj)",col="grey40",las=1,xlim=c(-6,6),pch=19)
	points(myx.wt,myy.wt,col="red",pch=19)
	points(myx.ko,myy.ko,col="blue",pch=19)
abline(v=0,col="red",lwd=1)
abline(v=log2(myratio),col="red",lwd=1,lty=2)	
abline(v=log2(1/myratio),col="red",lwd=1,lty=2)	
abline(h=-log10(myp),col="red",lwd=1)	
legend("topright",c(as.character(length(myx.wt)),as.character(length(myx.ko))),pch=19,col=c("red","blue"))


table(mysig.wt$conversion)
#~~~~~~~~~~~~~~~~~~~~~
# Final set objects
#~~~~~~~~~~~~~~~~~~~~~

dim(mysig_allmm)
```


# Volcanos of all mismatch types with same parameters as above


```{r volcano_allmm_wtsites, fig.width=10, fig.height=8}


candidates <- balanced_candidates

#~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Variation - PLotting as unadj p, different cutoff
#~~~~~~~~~~~~~~~~~~~~~~~~~~~

conv_tab <- table(candidates$conversion)
conv_tab[order(conv_tab)]

myconvs <- names(conv_tab)

# Set cutoff

myp <- 0.05
myratio <- 5

op <- par(mfrow = c(3,4))
for (i in 1:length(myconvs)) {
  
  myconv <- myconvs[i]
  myset <- candidates[candidates$conversion %in% myconv,]
  
  
  myx <- log2(myset$WT_KO_NaBH4_ratio)
	myy <- -log10(myset$Fisher_WT_KO_BH4.adj)
	myy[myy >= 10] <- 10
	myx[myx >= 6] <- 6
	myy[myy <= -10] <- -10
	myx[myx <= -6] <- -6

	# Get the sigs
	mysig.wt <- myset[myset$Fisher_WT_KO_BH4.adj <= myp & myset$WT_KO_NaBH4_ratio >= myratio,]
	#mysig.wt.ct <- mysig.wt[mysig.wt$conversion == "C>T",]
	myx.wt <- log2(mysig.wt$WT_KO_NaBH4_ratio)
	myy.wt <- -log10(mysig.wt$Fisher_WT_KO_BH4.adj)

	mysig.ko <- myset[myset$Fisher_WT_KO_BH4.adj <= myp & myset$WT_KO_NaBH4_ratio <= (1/myratio),]
	#mysig.ko <- myset[myset$Fisher_WT_KO_BH4 <= myp & myset$KO_WT_NaBH4_ratio >= myratio,]
	#mysig.ko.ct <- mysig.ko[mysig.ko$conversion == "C>T",]
	myx.ko <- log2(mysig.ko$WT_KO_NaBH4_ratio)
	myy.ko <- -log10(mysig.ko$Fisher_WT_KO_BH4.adj)

	myy.wt[myy.wt >= 10] <- 10
	myx.wt[myx.wt >= 6] <- 6
	myy.wt[myy.wt <= -10] <- -10
	myx.wt[myx.wt <= -6] <- -6

	myy.ko[myy.ko >= 10] <- 10
	myx.ko[myx.ko >= 6] <- 6
	myy.ko[myy.ko <= -10] <- -10
	myx.ko[myx.ko <= -6] <- -6  
	#~~~~~~~~~~~~~~~~~~~~~
	# Color coding by sig
	#~~~~~~~~~~~~~~~~~~~~~
	plot(myx,myy,main=paste0(myconv,
	" conversion\n(WT vs NAT10-/-)"),xlab="FC",ylab="-log10(P unadj)",col="grey40",las=1,xlim=c(-6,6),ylim=c(0,10),pch=19)
	points(myx.wt,myy.wt,col="red",pch=19)
	points(myx.ko,myy.ko,col="blue",pch=19)
	abline(v=0,col="red",lwd=1)
	abline(v=log2(myratio),col="red",lwd=1,lty=2)   
	abline(v=log2(1/myratio),col="red",lwd=1,lty=2) 
	abline(h=-log10(myp),col="red",lwd=1)   
	legend("top",c(as.character(length(myx.wt)),as.character(length(myx.ko))),pch=19,col=c("red","blue"))
}
par(op)

```


# All mismatch types, unadj p-val, balanced candidates

```{r volcano_CT_balanced_unadj, fig.width=10, fig.height=8}


candidates <- balanced_candidates

#~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Variation - PLotting as unadj p, different cutoff
#~~~~~~~~~~~~~~~~~~~~~~~~~~~

conv_tab <- table(candidates$conversion)
conv_tab[order(conv_tab)]

myconvs <- names(conv_tab)

# Set cutoff

myp <- 0.01
myratio <- 5

op <- par(mfrow = c(3,4))
for (i in 1:length(myconvs)) {
  
  myconv <- myconvs[i]
  myset <- candidates[candidates$conversion %in% myconv,]
  
  
  myx <- log2(myset$WT_KO_NaBH4_ratio)
	myy <- -log10(myset$Fisher_WT_KO_BH4)
	myy[myy >= 10] <- 10
	myx[myx >= 6] <- 6
	myy[myy <= -10] <- -10
	myx[myx <= -6] <- -6

	# Get the sigs
	mysig.wt <- myset[myset$Fisher_WT_KO_BH4 <= myp & myset$WT_KO_NaBH4_ratio >= myratio,]
	#mysig.wt.ct <- mysig.wt[mysig.wt$conversion == "C>T",]
	myx.wt <- log2(mysig.wt$WT_KO_NaBH4_ratio)
	myy.wt <- -log10(mysig.wt$Fisher_WT_KO_BH4)

	mysig.ko <- myset[myset$Fisher_WT_KO_BH4 <= myp & myset$WT_KO_NaBH4_ratio <= (1/myratio),]
	#mysig.ko <- myset[myset$Fisher_WT_KO_BH4 <= myp & myset$KO_WT_NaBH4_ratio >= myratio,]
	#mysig.ko.ct <- mysig.ko[mysig.ko$conversion == "C>T",]
	myx.ko <- log2(mysig.ko$WT_KO_NaBH4_ratio)
	myy.ko <- -log10(mysig.ko$Fisher_WT_KO_BH4)

	myy.wt[myy.wt >= 10] <- 10
	myx.wt[myx.wt >= 6] <- 6
	myy.wt[myy.wt <= -10] <- -10
	myx.wt[myx.wt <= -6] <- -6

	myy.ko[myy.ko >= 10] <- 10
	myx.ko[myx.ko >= 6] <- 6
	myy.ko[myy.ko <= -10] <- -10
	myx.ko[myx.ko <= -6] <- -6  
	#~~~~~~~~~~~~~~~~~~~~~
	# Color coding by sig
	#~~~~~~~~~~~~~~~~~~~~~
	plot(myx,myy,main=paste0(myconv,
	" conversion\n(WT vs NAT10-/-)"),xlab="FC",ylab="-log10(P unadj)",col="grey40",las=1,xlim=c(-6,6),ylim=c(0,10),pch=19)
	points(myx.wt,myy.wt,col="red",pch=19)
	points(myx.ko,myy.ko,col="blue",pch=19)
	abline(v=0,col="red",lwd=1)
	abline(v=log2(myratio),col="red",lwd=1,lty=2)   
	abline(v=log2(1/myratio),col="red",lwd=1,lty=2) 
	abline(h=-log10(myp),col="red",lwd=1)   
	legend("top",c(as.character(length(myx.wt)),as.character(length(myx.ko))),pch=19,col=c("red","blue"))
}
par(op)

```



# C>T, unadj p-val, balanced candidates

```{r volcano_allmm_balanced, fig.width=10, fig.height=8}


candidates <- balanced_candidates

#~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Variation - PLotting as unadj p, different cutoff
#~~~~~~~~~~~~~~~~~~~~~~~~~~~



# Set cutoff

myp <- 0.01
myratio <- 5


  myconv <- myconvs[i]
  myset <- candidates[candidates$conversion %in% "C>T",]
  
  
  myx <- log2(myset$WT_KO_NaBH4_ratio)
	myy <- -log10(myset$Fisher_WT_KO_BH4)
	myy[myy >= 10] <- 10
	myx[myx >= 6] <- 6
	myy[myy <= -10] <- -10
	myx[myx <= -6] <- -6

	# Get the sigs
	mysig.wt <- myset[myset$Fisher_WT_KO_BH4 <= myp & myset$WT_KO_NaBH4_ratio >= myratio,]
	#mysig.wt.ct <- mysig.wt[mysig.wt$conversion == "C>T",]
	myx.wt <- log2(mysig.wt$WT_KO_NaBH4_ratio)
	myy.wt <- -log10(mysig.wt$Fisher_WT_KO_BH4)

	mysig.ko <- myset[myset$Fisher_WT_KO_BH4 <= myp & myset$WT_KO_NaBH4_ratio <= (1/myratio),]
	#mysig.ko <- myset[myset$Fisher_WT_KO_BH4 <= myp & myset$KO_WT_NaBH4_ratio >= myratio,]
	#mysig.ko.ct <- mysig.ko[mysig.ko$conversion == "C>T",]
	myx.ko <- log2(mysig.ko$WT_KO_NaBH4_ratio)
	myy.ko <- -log10(mysig.ko$Fisher_WT_KO_BH4)

	myy.wt[myy.wt >= 10] <- 10
	myx.wt[myx.wt >= 6] <- 6
	myy.wt[myy.wt <= -10] <- -10
	myx.wt[myx.wt <= -6] <- -6

	myy.ko[myy.ko >= 10] <- 10
	myx.ko[myx.ko >= 6] <- 6
	myy.ko[myy.ko <= -10] <- -10
	myx.ko[myx.ko <= -6] <- -6  
	#~~~~~~~~~~~~~~~~~~~~~
	# Color coding by sig
	#~~~~~~~~~~~~~~~~~~~~~
	plot(myx,myy,main=paste0(myconv,
	" conversion\n(WT vs NAT10-/-)"),xlab="FC",ylab="-log10(P unadj)",col="grey40",las=1,xlim=c(-6,6),pch=19)
	points(myx.wt,myy.wt,col="red",pch=19)
	points(myx.ko,myy.ko,col="blue",pch=19)
	abline(v=0,col="red",lwd=1)
	abline(v=log2(myratio),col="red",lwd=1,lty=2)   
	abline(v=log2(1/myratio),col="red",lwd=1,lty=2) 
	abline(h=-log10(myp),col="red",lwd=1)   
	legend("top",c(as.character(length(myx.wt)),as.character(length(myx.ko))),pch=19,col=c("red","blue"))

```
## Using the balanced candidates set

```{r mm_barplot_balanced, fig.width=10, fig.height=5}


# With the wt candidates

mymm <- balanced_candidates

op <- par(mfrow = c(1,2))
wt_subs <- mymm$Sub_0
ko_subs <- mymm$Sub_1
barplot(table(factor(wt_subs,levels=seq(0,10,1))),ylab="Number of sites",xlab="Number of mismatches",main="Mismatches in WT\n(All mismatches, before statistical testing)",las=2)
barplot(table(factor(ko_subs,levels=seq(0,10,1))),ylab="Number of sites",xlab="Number of mismatches",main="Mismatches in NAT10-/-\n(All mismatches), before statistical testing",las=2)
par(op)

mymm <- balanced_candidates[balanced_candidates$conversion == "C>T",]

op <- par(mfrow = c(1,2))
wt_subs <- mymm$Sub_0
ko_subs <- mymm$Sub_1
barplot(table(factor(wt_subs,levels=seq(0,10,1))),ylab="Number of sites",xlab="Number of mismatches",main="Mismatches in WT\n(C>T, before statistical testing)",las=2,ylim=c(0,20000))
barplot(table(factor(ko_subs,levels=seq(0,10,1))),ylab="Number of sites",xlab="Number of mismatches",main="Mismatches in NAT10-/-\n(C>T, before statistical testing",las=2,ylim=c(0,20000))
par(op)

```

# Barplot of N values from the Volcano

```{r balanced_volcano_Ns, fig.width=5, fig.height=5}

myp <- 0.01
myratio <- 5

myset <- balanced_candidates
mysig.wt <- myset[myset$Fisher_WT_KO_BH4 <= myp & myset$WT_KO_NaBH4_ratio >= myratio,]
mysig.ko <- myset[myset$Fisher_WT_KO_BH4 <= myp & myset$WT_KO_NaBH4_ratio <= (1/myratio),]


tobar <- rbind(table(factor(mysig.wt$conversion,levels=myconvs)),
                     table(factor(mysig.ko$conversion,levels=myconvs)))
barplot(tobar,beside=T,col=c("red","blue"),las=1,main="Mismatch calls (All MM types)\nUnadjusted p <= 0.01, Ratio >= 5",legend.text = c("Up in WT","Up in NAT10-/-"),ylab="Number of sites",xlab="Mismatch type",ylim=c(0,8000),las=2)

```


# Session info

```{r}
sessionInfo()
```